### PR TITLE
Add a dev mode that allows for local storing of config files and logs

### LINKED ISSUE #1162: FR: Local config mode for development.
When I start *vorta* from my local python environment, it will always use the global settings and config files. I feel very uneasy when sharing the config files of my real backup system with development versions and branches. This can get especially problematic when a new version uses a different config setup or file format. I also see a problem for manual (and possibly automatic) testing.

A solution would be the option to call *vorta* with an command-line argument like `--local` that starts it in a local config mode. In this mode vorta would not use the user wide location for config files but one nearby *(e.g. in the cwd)*.

### PR DESCRIPTION
**Pull request by ratchek**
### Description
Allows vorta to be called with a command-line flag that will make it use a directory in the project tree to store all the settings, logs, and cache.
Also allows for a custom directory path allowing for multiple "configuration" folders at once.

### Related Issue
Fixes #1162

### Motivation and Context
This can be used to prevent the vorta instance that a developer is working on from accessing the configuration files that they have set up for their personal backups.

### How Has This Been Tested?
Just been tested in the terminal by me.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:

- [x] I have read the [CONTRIBUTING](https://vorta.borgbase.com/contributing/) guide.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

*I provide my contribution under the terms of the [license](./../../LICENSE.txt) of this repository and I affirm the [Developer Certificate of Origin][dco].*

[dco]: https://developercertificate.org/

### CODE REVIEW

**Comment by real-yfprojects** [comment]
Note how you can write `Fixed #1162` to close the issue when merging the PR.

Fixes #1162.

File: src/vorta/__main__.py
Comment (real-yfprojects) [inline src/vorta/__main__.py:8]: :+1: Nice commentary!

File: src/vorta/__main__.py
Comment (real-yfprojects) [inline src/vorta/__main__.py:52]:
```suggestion
    want_development = getattr(args, 'development', False)
```

I think an exception will be raised if the attribute doesn't exist and no default value is supplied.

File: src/vorta/__main__.py
Comment (real-yfprojects) [inline src/vorta/__main__.py:93]: It seems like you either forgot to remove these lines or reimplement the exception handling.

File: src/vorta/__main__.py
Comment (real-yfprojects) [inline src/vorta/__main__.py:69]:
```suggestion
        if want_development != DEFAULT_DIR_FLAG:
            config.DEV_MODE_DIR = Path(want_development)
```

File: src/vorta/utils.py
Comment (real-yfprojects) [inline src/vorta/utils.py:25]:
```suggestion
DEFAULT_DIR_FLAG = object()
```

This is an actually unique value. The user could theoretically supply the path `USE_DEFAULT_DIR`.

File: src/vorta/__main__.py
Comment (real-yfprojects) [inline src/vorta/__main__.py:90]: Can this be moved into `vorta.config`? And can you then try to reduce redundancy that is duplicated code?

**Review by real-yfprojects — COMMENTED** [review]
Thank you very much for taking on this issue. That will be a very useful development feature.

**Comment by ratchek** [comment]
> There are some commits in the branch diff that do not belong to this PR. Can you fix that?

Could you elaborate on that please? This is the first time I'm using git in a collaborative manner. I take it you mean that there are commits arising from me merging in work that other people did? How do I go about fixing that?

File: src/vorta/utils.py
Comment (ratchek) [reply src/vorta/utils.py:25]: Ah, that's clever. I wasn't sure how to go about this and just went for a string that was unlikely to be used but still somewhat informative.

File: src/vorta/__main__.py
Comment (ratchek) [reply src/vorta/__main__.py:90]: That's what I initially planned on doing, but the result was a circular dependency I wasn't quite sure how to fix:

https://github.com/borgbase/vorta/issues/1162#issuecomment-1445214054

**Comment by real-yfprojects** [comment]
> I take it you mean that there are commits arising from me merging in work that other people did? How do I go about fixing that?

Yes, usually these commits aren't shown in the branch diff. Here they are because you somehow rewrote the commits making you the committer. I managed to fix this by running

```bash
git rebase -i master
```
And dropping the commits in question since they are already included in the master branch. I forced pushed the fixed branch, so you will have to update your local copy of it.

File: src/vorta/__main__.py
Comment (real-yfprojects) [reply src/vorta/__main__.py:90]: This is what I came up with. However it isn't perfect since our config dirs aren't real constants any more and can't be imported with `from vorta.config import LOG_DIR` because they won't be updated when imported like that.  Instead one always has to use `from vorta import config`.

```py
SETTINGS_DIR = None
LOG_DIR = None
CACHE_DIR = None
TEMP_DIR = None
PROFILE_BOOTSTRAP_FILE = None


def default_dev_dir():
    return Path(__file__).parent.parent.parent / '.dev_config'


def init_from_platformdirs():
    dirs = platformdirs.PlatformDirs(APP_NAME, APP_AUTHOR)
    init(dirs.user_data_path, dirs.user_log_path, dirs.user_cache_path, dirs.user_cache_path / 'temp', Path.home())


def init_dev_mode(dir):
    init(dir / 'settings', dir / 'logs', dir / 'cache', dir / 'temp', dir)


def init(settings, logs, cache, temp, bootstrap):
    global SETTINGS_DIR
    global LOG_DIR
    global CACHE_DIR
    global TEMP_DIR
    global PROFILE_BOOTSTRAP_FILE
    SETTINGS_DIR = settings
    LOG_DIR = logs
    CACHE_DIR = cache
    TEMP_DIR = temp
    PROFILE_BOOTSTRAP_FILE = bootstrap / '.vorta-init.json'
    ensure_dirs()


def ensure_dirs():
    # ensure directories exist
    for dir in (SETTINGS_DIR, LOG_DIR, CACHE_DIR, TEMP_DIR):
        dir.mkdir(parents=True, exist_ok=True)
```

Comment (ratchek) [reply src/vorta/__main__.py:90]: Ok, so I'm gonna do this and then search through the codebase for any imports from config and modify accordingly. Sounds good?

Comment (ratchek) [reply src/vorta/__main__.py:90]: Ok, so I'm gonna do this and then search through the codebase for any imports from config and modify accordingly. Sounds good?

**Comment by real-yfprojects** [comment]
This is how I do it:

Keep master branch in sync with upstream
```
git checkout master
git pull upstream
```

Now you can either merge master into the feature branch. This what many people do. Its the easiest method:
```
git checkout feature-branch
git merge master
```

However I prefer keeping clean and linear git histories. I therefore rebase my feature branch onto master:

```
git checkout feature-branch
git rebase master
```

The rebase will also remove all merge commits of master that occur after the diverging.

I would be interested how you managed to get branch you had before. My only guess was that you might have merged `origin/master` or `upstream/master` and than rebased onto `master` which wasn't in sync with the remote branches.

**Comment by ratchek** [comment]
Ok, that's pretty much exactly what I do most of the time. The problem is, I think whenever I get merge conflicts (and I think that's where I git those weird results last time - trying to resolve merge conflicts).

Sidenote:
When I resolve merge conflicts and commit the changes, what should I be putting in the commit message? Just a generic "merge changes from upstream"? That seems redundant and unhelpful, but I don't know what the alternative might be.

**Comment by ratchek** [comment]
Don't merge this yet please. I didn't have a chance to test it out, because the new version with Qt6 is giving me issues (even when doing a fresh install from the main repo) and I need to fix that first.

File: src/vorta/__main__.py
Comment (ratchek) [reply src/vorta/__main__.py]: > Ok, so I'm gonna do this and then search through the codebase for any imports from config and modify accordingly. Sounds good?

I modified all the imports except the one in `./package/vorta.spec` on line 7. That one doesn't use any of the variables that are changing. Only APP_NAME and APP_ID_DARWIN, and those are both just assigned as constants. Should I leave it like that, or should I also import the whole module for consistency's sake?

**Comment by real-yfprojects** [comment]
> Ok, that's pretty much exactly what I do most of the time. The problem is, I think whenever I get merge conflicts (and I think that's where I git those weird results last time - trying to resolve merge conflicts).

Merge conflicts are usually resolved by the merge commit itself without rewriting any previous commits. Only few commands rewrite git history.

> When I resolve merge conflicts and commit the changes, what should I be putting in the commit message? Just a generic "merge changes from upstream"?

That's fine. The merge commit will have the commits with helpful descriptions as a parent. Mostly we squash PRs anyway.

File: src/vorta/application.py
Comment (real-yfprojects) [inline src/vorta/application.py:26]: Typo

File: src/vorta/application.py
Comment (real-yfprojects) [inline src/vorta/application.py:28]: Typo

File: src/vorta/application.py
Comment (real-yfprojects) [inline src/vorta/application.py:30]: I might be better to call `init_from_platformdirs` in `config` (every time) and override it in `__main__` when needed.

File: src/vorta/application.py
Comment (real-yfprojects) [inline src/vorta/application.py:267]: When is `config.PROFILE_BOOTSTRAP_FILE` evaluated? 

I think it is evaluated when the module is imported. Therefore changes to `config` aren't respected.

File: src/vorta/application.py
Comment (ratchek) [reply src/vorta/application.py:26]: Sorry, what does this mean? I can't find the typo.

File: src/vorta/application.py
Comment (ratchek) [reply src/vorta/application.py:267]: I'm not sure what to do about this. Could you point me in the right direction?

File: src/vorta/application.py
Comment (real-yfprojects) [reply src/vorta/application.py:26]: It sais `withoug` instead of `without`.

File: src/vorta/application.py
Comment (real-yfprojects) [reply src/vorta/application.py:267]:
```suggestion
    def bootstrap_profile(self, bootstrap_file=None):
        bootstrap_file = bootstrap_file or config.PROFILE_BOOTSTRAP_FILE
```

File: src/vorta/application.py
Comment (ratchek) [reply src/vorta/application.py:30]: You mean calling `init_from_platformdirs` in `application`? Or calling that method in `config`  so that it runs every time the module is loaded? Wouldn't the latter risk overwriting the variables from time to time as a side effect?

Comment (real-yfprojects) [reply src/vorta/application.py:30]: > Or calling that method in config so that it runs every time the module is loaded?

That's what I was thinking of since it ensures the config values contain valid values.

> Wouldn't the latter risk overwriting the variables from time to time as a side effect?

Python modules are only loaded once (and cached) in a run afaik.

Comment (ratchek) [reply src/vorta/application.py:30]: Ohhh. I see. Thanks! I'll do that.

**Comment by ratchek** [comment]
So, I just realized that now there's this issue, where everything works fine, unless you call `vorta` with the `--development` flag and pass it a custom directory using relative paths. So for example, if you do `vorta --development ./dev-mode`. In that particular scenario, the link to logs in the misc tab doesn't work. The logs exist and they're in the correct folder, but the link doesn't work.
I would suspect the call to `config` on line 37 in `views/misc_tab.py` is happening before the actual modification of the config values, however if you supply the full path when you're running the program (for example `vorta --development /home/username/path/to/project/folder/dev-mode`), then everything works fine. Is it possible that the string interpolation is not unpacking the full path correctly?

**Comment by ratchek** [comment]
> Is it possible that the string interpolation is not unpacking the full path correctly?

Yup. That was it. Fixed it. Should be good now.

**Review by real-yfprojects — COMMENTED** [review]
Nice! Can you add type hints and docstrings?

File: src/vorta/utils.py
Comment (real-yfprojects) [inline src/vorta/utils.py:373]: Can you specify a `metavar` so that it doesn't say 

```
usage: vorta [-h] [--version] [--daemonize] [--create PROFILE]
             [--development [DEVELOPMENT]]
```

when running `vorta --help`?

**Review by real-yfprojects — APPROVED** [review]

**Comment by real-yfprojects** [comment]
Nice work, thanks!

**Comment by ratchek** [comment]
> Nice work, thanks!

Thank you so so much for guiding me through this and being so patient.

**Comment by m3nu** [comment]
That's fine. I see you mostly changed `XXX` to `config.XXX`. Reminds me a bit of Django now.

Updating the docs would be important too. So people actually use this feature as intended.

**Review by m3nu — APPROVED** [review]

**Comment by real-yfprojects** [comment]
> Updating the docs would be important too. So people actually use this feature as intended.

Follow up issue: borgbase/vorta.borgbase.com#42

**Pull request merged**
